### PR TITLE
Resolves #3151 plus redo and improve subject mapping destination validation and error handling

### LINKED ISSUE #3151: Mappings cannot have spaces in mapping functions
## Defect

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `nats-server` and affected client libraries used:

a52f12613e6eb53097d6d563d9b9681b88a037cb

#### OS/Container environment:

Linux EL8

#### Steps or code to reproduce the issue:

Use configuration:

```
mappings = {
  "in.registration.*.>": "registration.{{ partition(5, 1) }}.{{wildcard(1)}}.>"
}
```

```
$ nats-server --config /tmp/server.conf
nats-server: /tmp/server.conf:2:4: Error adding mapping for "in.registration.*.>" to "registration.{{partition(5, 1)}}.{{wildcard(1)}}.>" : invalid subject
```

But if we adjust the subject mapping like this (remove all spaces):

```
mappings = {
  "in.registration.*.>": "registration.{{partition(5,1)}}.{{wildcard(1)}}.>"
}
```

It works.

This is because https://github.com/nats-io/nats-server/blob/a52f12613e6eb53097d6d563d9b9681b88a037cb/server/accounts.go#L618-L621  is checking the transform subject for validity before any parsing is done, I think probably we should remove that check.  Spaces in these mappings greatly improves readability and we specifically supported them in the code when implementing it.

### PR DESCRIPTION
**Pull request by jnmoyne**
- [X] Link to issue, e.g. `Resolves #NNN`
 - [X] Documentation added (if applicable)
 - [X] Tests added
 - [X] Branch rebased on top of current main (`git pull --rebase origin main`)
 - [X] Changes squashed to a single commit (described [here](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
 - [ ] Build is green in Travis CI
 - [X] You have certified that the contribution is your original work and that you license the work to the project under the [Apache 2 license](https://github.com/nats-io/nats-server/blob/main/LICENSE)

Resolves #3151

### Changes proposed in this pull request:

1: Allows spaces to be used inside {{}} subject mapping functions:

2: Rework and improve mapping destinations validation and error handling with much more helpful error messages, e.g.:

* Error adding mapping for "foo.*.*" to "bar.{{wildcard(1)}}" : invalid mapping destination: not using all of the token wildcard(s) in bar.{{wildcard(1)}}

* Error adding mapping for "myservice.request.*" to "myservice.request.{{ partition(10) }}.{{wildcard(1)}}" : invalid mapping destination: not enough arguments passed to the function in {{ partition(10) }}

* Error adding mapping for "myservice.request.*" to "myservice.request.{{ partition(10,2) }}.{{wildcard(1)}}" : invalid mapping destination: wildcard index out of range in {{ partition(10,2) }}: [2]

* Error adding mapping for "myservice.request.*" to "myservice.request.{{ partition(10,1) }}.{{wildcard()}}" : invalid mapping destination: not enough arguments passed to the function in {{wildcard()}}

* Error adding mapping for "myservice.request.*" to "myservice.request.{{ xxxpartition(10,1) }}.{{wildcard(1)}}" : invalid mapping destination: unknown function in {{ xxxpartition(10,1) }}

* Error adding mapping for "myservice.request.*" to "myservice. request.{{ xxxpartition(10,1) }}.{{wildcard(1)}}" : invalid mapping destination: invalid subject

/cc @nats-io/core

### CODE REVIEW

**Review by ripienaar — APPROVED** [review]
I like it, thanks, but lets see Ivan and Dereks take on it. I mainly care for the UX part of things - this will help a lot.

File: server/accounts.go [hunk 1]
```diff
@@ -4189,33 +4190,41 @@ func placeHolderIndex(token string) ([]int, int32, error) {
 			return []int{tp}, -1, nil
 		}
 
-		// New 'moustache' style mapping
+		// New 'mustache' style mapping
 		// wildcard(wildcard token index) (equivalent to $)
 		args := getMappingFunctionArgs(wildcardMappingFunctionRegEx, token)
 		if args != nil {
+			if len(args) == 1 && args[0] == "" {
```
Comment (kozlovic) [inline server/accounts.go:4197]: nit: Derek likes to use `_EMPTY_` instead of `""`

**Review by kozlovic — CHANGES_REQUESTED** [review]
Minor comments about coding style.

File: server/accounts.go [hunk 2]
```diff
@@ -4287,8 +4296,8 @@ func newTransform(src, dest string) (*transform, error) {
 				dtpinb = append(dtpinb, -1)
 			}
 		}
-		if nphs < npwcs {
-			return nil, ErrBadSubjectMappingDestination
+		if nphs < npwcs { // not all wildcards are being used in the destination
```
[the remark this answers is not in the document]
Comment (jnmoyne) [reply server/accounts.go]: done

File: server/accounts.go (hunk 1 above)
Comment (jnmoyne) [reply server/accounts.go]: Done

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**